File: Sources/Droid/Packages/swift/NativeTextWatcher.swift
//
//  TextWatcher.swift
//  Droid
//
//  Created by Mihael Isaev on 27.01.2022.
//

import DroidFoundation
import FoundationEssentials

extension SwiftPackage {
    public class TextWatcherClass: JClassName, @unchecked Sendable {}
    
    public var TextWatcher: TextWatcherClass { .init(parent: self, name: "NativeTextWatcher") }
}

// var _nativeTextWatchers: [NativeTextWatcher] = []

// class NativeTextWatcher: @unchecked Sendable, JObjectable {
//     init (_ environment: JEnvironment, _ context: JObjectReference) {
//         super.init(environment, context, classes: [.swift.TextWatcher], args: [])
//         _nativeTextWatchers.append(self)
//     }
    
//     required init(_ environment: JEnvironment, _ ref: JClassReference, _ object: jobject) {
//         super.init(environment, ref, object)
//         _nativeTextWatchers.append(self)
//     }
    
//     deinit {
//         _nativeTextWatchers.removeAll(where: { $0 === self })
//     }
    
//     var _beforeTextChanged: (String, Int, Int, Int) -> Void = { _,_,_,_ in }
    
//     public func beforeTextChanged(_ handler: @escaping (_ text: String, _ start: Int, _ count: Int, _ after: Int) -> Void) -> Self {
//         _beforeTextChanged = handler
//         return self
//     }
    
//     var _onTextChanged: (String, Int, Int, Int) -> Void = { _,_,_,_ in }
    
//     public func onTextChanged(_ handler: @escaping (_ text: String, _ start: Int, _ before: Int, _ count: Int) -> Void) -> Self {
//         _onTextChanged = handler
//         return self
//     }
    
//     var _afterTextChanged: (String) -> Void = { _ in }
    
//     public func afterTextChanged(_ handler: @escaping (_ text: String) -> Void) -> Self {
//         _afterTextChanged = handler
//         return self
//     }
// }

// @_cdecl("Java_swift_NativeTextWatcher_beforeTextChanged")
// public func text_watcher_beforeTextChanged(env: UnsafeMutablePointer<JNIEnv?>, callerClassObject: jobject, spannableStringBuilder: jobject, start: jint, count: jint, after: jint) {
//     let env = JEnvironment(env)
//     let sss = SpannableStringBuilder(env, JClassReference(env, [.android.text.SpannableStringBuilder]), spannableStringBuilder)
//     _nativeTextWatchers.first(where: { env.isSameObject($0.classObject, callerClassObject) })?._beforeTextChanged(sss.toString(), Int(start), Int(count), Int(after))
// }

// @_cdecl("Java_swift_NativeTextWatcher_onTextChanged")
// public func text_watcher_onTextChanged(env: UnsafeMutablePointer<JNIEnv?>, callerClassObject: jobject, spannableStringBuilder: jobject, start: jint, before: jint, count: jint) {
//     let env = JEnvironment(env)
//     let sss = SpannableStringBuilder(env, JClassReference(env, [.android.text.SpannableStringBuilder]), spannableStringBuilder)
//     _nativeTextWatchers.first(where: { env.isSameObject($0.classObject, callerClassObject) })?._onTextChanged(sss.toString(), Int(start), Int(before), Int(count))
// }

// @_cdecl("Java_swift_NativeTextWatcher_afterTextChanged")
// public func text_watcher_afterTextChanged(env: UnsafeMutablePointer<JNIEnv?>, callerClassObject: jobject, spannableStringBuilder: jobject) {
//     let env = JEnvironment(env)
//     let sss = SpannableStringBuilder(env, JClassReference(env, [.android.text.SpannableStringBuilder]), spannableStringBuilder)
//     _nativeTextWatchers.first(where: { env.isSameObject($0.classObject, callerClassObject) })?._afterTextChanged(sss.toString())
// }

File: appkit/src/main/java/stream/swift/droid/appkit/watchers/NativeTextWatcher.kt
package stream.swift.droid.appkit.watchers

import android.text.Editable
import android.text.TextWatcher

class NativeTextWatcher(private val uniqueId: Int): TextWatcher {
    override fun beforeTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {
        if (p0 != null) {
            beforeTextChangedExtended(uniqueId, p0.toString(), p1, p2, p3)
        } else {
            beforeTextChanged(uniqueId, p1, p2, p3)
        }
    }
    override fun onTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {
        if (p0 != null) {
            onTextChangedExtended(uniqueId, p0.toString(), p1, p2, p3)
        } else {
            onTextChanged(uniqueId, p1, p2, p3)
        }
    }
    override fun afterTextChanged(p0: Editable?) {
        if (p0 != null) {
            afterTextChangedExtended(uniqueId, p0.toString())
        } else {
            afterTextChanged(uniqueId)
        }
    }

    private external fun beforeTextChanged(uniqueId: Int, p1: Int, p2: Int, p3: Int)
    private external fun beforeTextChangedExtended(uniqueId: Int, p0: String, p1: Int, p2: Int, p3: Int)
    private external fun onTextChanged(uniqueId: Int, p1: Int, p2: Int, p3: Int)
    private external fun onTextChangedExtended(uniqueId: Int, p0: String, p1: Int, p2: Int, p3: Int)
    private external fun afterTextChanged(uniqueId: Int)
    private external fun afterTextChangedExtended(uniqueId: Int, p0: String)
}